File: ml/logic/cnf.py
from copy import deepcopy
from collections import defaultdict


class CNF(object):
    def __init__(self, problem):
        self.clauses = []
        self.lit_to_clause_indices = defaultdict(lambda: set())
        self.num_vars = problem.num_vars

        for gate in problem.gates:
            for clause in gate.cnf_clauses():
                self.clauses.append(set(clause))

        self.reindex_lit_to_clause()

    @property
    def num_clauses(self):
        return len(self.clauses)

    def __repr__(self):
        s = ''
        for i, clause in enumerate(self.clauses):
            s += f'{list(sorted(clause))}'
            if i != len(self.clauses) - 1:
                s += '\n'
        return s

    def reindex_lit_to_clause(self):
        self.lit_to_clause_indices = defaultdict(lambda: set())
        clause_idx = 0
        new_clauses = []

        for clause in self.clauses:
            if clause is None:
                continue  # Skip

            clause_is_auto_sat = False
            for lit in clause:
                if (-lit) in clause:
                    clause_is_auto_sat = True
                    break

            if clause_is_auto_sat:
                continue  # Skip

            for lit in clause:
                self.lit_to_clause_indices[lit].add(clause_idx)

            new_clauses.append(clause)
            clause_idx += 1

        self.clauses = new_clauses

    def simplify(self, assumptions):
        cnf = deepcopy(self)
        assumptions = deepcopy(assumptions)
        queue = []

        for lit, val in assumptions.items():
            assert lit > 0
            queue.append((lit, val))
            queue.append((-lit, 1 - val))

        while len(queue) > 0:
            lit, val = queue.pop()
            referenced_clauses = cnf.lit_to_clause_indices[lit]
            if val:
                for clause_idx in referenced_clauses:
                    # All referenced clauses are automatically SAT
                    cnf.clauses[clause_idx] = None  # Remove clause
            else:
                for clause_idx in referenced_clauses:
                    clause = cnf.clauses[clause_idx]
                    if clause is None:
                        continue

                    # If this is the last literal in the clause, UNSAT
                    if len(clause) == 1:
                        return None

                    # Remove literal from clause, since it is 0 / false
                    clause.discard(lit)

                    # If clause now has only 1 lit, add it to assumptions
                    if len(clause) == 1:
                        last_lit = next(iter(clause))

                        if last_lit in assumptions and assumptions[last_lit] != 1:
                            return None  # UNSAT because we need last_lit = 1
                        if (-last_lit) in assumptions and assumptions[-last_lit] != 0:
                            return None  # UNSAT, need -last_lit = 0 --> last_lit = 1

                        queue.append((last_lit, 1))
                        queue.append((-last_lit, 0))
                        assumptions[last_lit] = 1
                        assumptions[-last_lit] = 0
                        clause = None  # Remove clause (because now it is SAT)

                    cnf.clauses[clause_idx] = clause

        cnf.reindex_lit_to_clause()
        return cnf

    def is_sat(self, assignments):
        for clause in self.clauses:
            clause_is_sat = any(assignments[lit] for lit in clause)
            if not clause_is_sat:
                return False
        return True

File: src/core/cnf.cpp
/*
 * Copyright (c) 2022 Authors:
 *   - Trevor Phillips <trevphil3@gmail.com>
 *
 * Distributed under the CC BY-NC-SA 4.0 license
 * (See accompanying file LICENSE.md).
 */

#include "core/cnf.hpp"

#include <assert.h>

#include <algorithm>
#include <fstream>
#include <iterator>
#include <regex>
#include <sstream>
#include <stdexcept>
#include <utility>

std::string trim(const std::string &s) {
  std::regex e("^\\s+|\\s+$");
  return std::regex_replace(s, e, "");
}

std::vector<std::string> splitByWhitespace(const std::string &s) {
  std::istringstream buffer(s);
  std::vector<std::string> ret;
  std::copy(std::istream_iterator<std::string>(buffer),
            std::istream_iterator<std::string>(), std::back_inserter(ret));
  return ret;
}

namespace preimage {

struct Simplification {
 public:
  Simplification(const CNF &cnf, std::unordered_map<int, bool> &assignments) {
    std::vector<std::pair<int, bool>> q;
    std::vector<std::set<int>> tmp_clauses = cnf.clauses;
    std::unordered_map<int, std::set<int>> lit2clauses;

    // Mapping from: literal --> indices of clauses which use that literal
    for (int c = 0; c < cnf.num_clauses; ++c) {
      for (int lit : tmp_clauses[c]) lit2clauses[lit].insert(c);
    }

    // Queue will contain literals for which we KNOW the assignment
    for (const auto &itr : assignments) {
      if (itr.first <= 0) {
        char err_msg[128];
        snprintf(err_msg, 128,
                 "CNF assignments should be 1-indexed and not negated (got %d)",
                 itr.first);
        throw std::invalid_argument(err_msg);
      }
      q.push_back({itr.first, itr.second});
      q.push_back({-itr.first, !itr.second});
    }

    while (q.size() > 0) {
      // Pop from queue and cache the known assignment
      const int lit = q.back().first;
      const bool val = q.back().second;
      q.pop_back();
      assignments[lit] = val;
      assignments[-lit] = !val;

      const std::set<int> &clause_indices = lit2clauses[lit];
      if (val) {
        // All referenced clauses are automatically SAT
        for (int clause_idx : clause_indices) tmp_clauses[clause_idx] = {};
      } else {
        for (int clause_idx : clause_indices) {
          if (tmp_clauses[clause_idx].size() == 0) continue;
          if (tmp_clauses[clause_idx].size() == 1) {
            // If this is the last literal in the clause, UNSAT
            throw std::runtime_error("Problem found to be UNSAT during simplification");
          }
          // Remove literal from clause, since it is 0 / false
          tmp_clauses[clause_idx].erase(lit);
          if (tmp_clauses[clause_idx].size() == 1) {
            const int last_lit = *(tmp_clauses[clause_idx].begin());
            if (assignments.count(last_lit) && !assignments[last_lit]) {
              throw std::runtime_error("Problem found to be UNSAT during simplification");
            }
            if (assignments.count(-last_lit) && assignments[-last_lit]) {
              throw std::runtime_error("Problem found to be UNSAT during simplification");
            }
            q.push_back({last_lit, true});
            q.push_back({-last_lit, false});
            tmp_clauses[clause_idx] = {};
          }
        }
      }
    }

    std::unordered_map<int, int> lit_original_to_simplified;
    std::vector<std::set<int>> simplified_clauses;
    int k = 1;

    for (const std::set<int> &orig_clause : tmp_clauses) {
      if (orig_clause.size() == 0) continue;  // Automatically SAT
      std::set<int> simplified_clause;
      for (int orig_lit : orig_clause) {
        const int positive_lit = std::abs(orig_lit);
        if (lit_original_to_simplified.count(positive_lit) == 0) {
          lit_original_to_simplified[positive_lit] = k++;
        }
        int new_lit = lit_original_to_simplified[positive_lit];
        new_lit *= (orig_lit == positive_lit ? 1 : -1);
        simplified_clause.insert(new_lit);
      }
      simplified_clauses.push_back(simplified_clause);
    }

    original_cnf = cnf;
    simplified_cnf = CNF(simplified_clauses, k - 1);

    // We COULD use less memory by storing in an array of length "k"
    lit_simplified_to_original = {};
    for (const auto &itr : lit_original_to_simplified) {
      lit_simplified_to_original[itr.second] = itr.first;
    }
  }

  CNF original_cnf, simplified_cnf;
  std::unordered_map<int, int> lit_simplified_to_original;
};

CNF::CNF() : num_vars(0), num_clauses(0), clauses({}) {}

CNF::CNF(const std::vector<LogicGate> &gates) {
  num_vars = 0;
  clauses = {};
  for (const LogicGate &gate : gates) {
    const std::vector<std::vector<int>> gate_clauses = gate.cnf();
    for (const std::vector<int> &gc : gate_clauses) {
      clauses.push_back(std::set<int>(gc.begin(), gc.end()));
      for (int lit : gc) num_vars = std::max(num_vars, std::abs(lit));
    }
  }
  num_clauses = static_cast<int>(clauses.size());
}

CNF::CNF(const std::vector<std::set<int>> &cls, int n_var) {
  num_vars = n_var;
  clauses = cls;
  num_clauses = static_cast<int>(clauses.size());
}

int CNF::numSatClauses(const std::unordered_map<int, bool> &assignments) const {
  int num_sat = 0;

  for (const auto &clause : clauses) {
    for (int lit : clause) {
      bool lit_val;
      if (assignments.count(lit))
        lit_val = assignments.at(lit);
      else if (assignments.count(-lit))
        lit_val = !assignments.at(-lit);
      else {
        char err_msg[128];
        snprintf(err_msg, 128, "CNF is missing assignment for literal %d", lit);
        throw std::out_of_range(err_msg);
      }

      if (lit_val) {
        num_sat++;
        break;
      }
    }
  }

  return num_sat;
}

double CNF::approximationRatio(const std::unordered_map<int, bool> &assignments) const {
  if (num_clauses == 0) return 1.0;
  return numSatClauses(assignments) / static_cast<double>(num_clauses);
}

void CNF::toFile(const std::string &filename) const {
  std::ofstream cnf_file(filename);
  if (!cnf_file.is_open()) {
    char err_msg[256];
    snprintf(err_msg, 256, "Unable to open in write mode: %s", filename.c_str());
    throw std::invalid_argument(err_msg);
  }

  cnf_file << "p cnf " << num_vars << " " << num_clauses << "\n";
  for (const std::set<int> &clause : clauses) {
    for (int var : clause) cnf_file << var << " ";
    cnf_file << "0\n";
  }

  cnf_file.close();
}

CNF CNF::fromFile(const std::string &filename) {
  std::ifstream cnf_file(filename);
  if (!cnf_file.is_open()) {
    char err_msg[256];
    snprintf(err_msg, 256, "Unable to open in read mode: %s", filename.c_str());
    throw std::invalid_argument(err_msg);
  }

  std::string line;
  int num_vars = -1;
  int num_clauses = -1;
  std::vector<std::set<int>> clauses;
  std::vector<std::string> parts;

  while (std::getline(cnf_file, line)) {
    line = trim(line);
    if (line.size() == 0 || line[0] == '#') continue;
    if (line.rfind("p cnf ", 0) == 0) {
      line = trim(line.substr(6));
      parts = splitByWhitespace(line);
      num_vars = std::stoi(parts[0]);
      num_clauses = std::stoi(parts[1]);
    } else {
      parts = splitByWhitespace(line);
      std::set<int> clause;
      for (const std::string &s : parts) {
        const int lit = std::stoi(s);
        if (lit != 0) clause.insert(lit);
      }
      clauses.push_back(clause);
    }
  }
  cnf_file.close();

  if (num_vars == -1 || num_clauses == -1) {
    throw std::runtime_error("Did not read `p cnf N M`, ensure DIMACS header exists");
  }

  return CNF(clauses, num_vars);
}

CNF CNF::simplify(std::unordered_map<int, bool> &assignments,
                  std::unordered_map<int, int> &lit_new_to_old) const {
  const Simplification simplification(*this, assignments);
  lit_new_to_old = simplification.lit_simplified_to_original;
  return simplification.simplified_cnf;
}

void CNF::toMIS(const std::string &) const {
  throw std::runtime_error("Function not yet implemented.");  // TODO(Trevor)
}

void CNF::toGraphColoring(const std::string &) const {
  throw std::runtime_error("Function not yet implemented.");  // TODO(Trevor)
}

}  // end namespace preimage
